### PR TITLE
Synthetic AuthService annotation config

### PR DESCRIPTION
**Pull request by AliceProxy**
Signed-off-by: AliceProxy <alicewasko@datawire.io>

## Description
This PR has multiple changes for instances of edge-stack and is best reviewed commit-by-commit. 

The first 6 commits are a rebase necessary to support the remaining commits that drop the following commit hashes from [Pull #4126](https://github.com/emissary-ingress/emissary/pull/4216) which cause various issues when pulled into the edge-stack codebase.
[b494d9](https://github.com/emissary-ingress/emissary/pull/4216/commits/b494d995b7d040be591ed61556630870baf11aee)
[09ecd8](https://github.com/emissary-ingress/emissary/pull/4216/commits/09ecd8e2dadff202a5349f1a1948b920f0a38718)
[dfbdc9](https://github.com/emissary-ingress/emissary/pull/4216/commits/dfbdc99af618017882ae7a231106491cb866ec26)
[4fae46](https://github.com/emissary-ingress/emissary/pull/4216/commits/4fae46776d028bb2ce204f729196985e805166cd)
[adbb2b](https://github.com/emissary-ingress/emissary/pull/4216/commits/adbb2b73e193584c88e1ccf064aa3a6bb66de2a4)

The first commit immediately after that simply adds Filters/Filterpolicies to the interesting types so that additional work on the edge-stack side can simply read the config for these resources from the snapshot for improved performance.
 
The remaining commits center around logic for the synthetic AuthService that is injected only when running edge-stack and no other valid AuthServices can be found. They add checks for AuthServices created by annotation config and also prevent tests that create custom AuthServices from running in edge-stack since custom AuthServices are not supported in edge-stack.

## Related Issues
List related issues.

## Testing
Automated tests for the synthetic AuthService already exist within the Emissary CI to ensure that it is injected successfully, these changes are mainly setup work for edge-stack and will be covered by edge-stack's tests.

## Checklist

 - ~~[ ] I made sure to update `CHANGELOG.md`.~~
   I did not add a changelog entry for this PR since all of these changes target edge-stack and will be covered by an addition to the edge-stack changelog.
 
 - [x] This is unlikely to impact how Ambassador performs at scale.
 
None of these changes will impact emissary-ingress since they only run when using edge-stack. Adding the Filters/Filterpolicies to the snapshot itself will not have much of an impact how Ambassador performs at scale, but it is a necessary change that will enable us to actually improve performance for loading filters and processing config changes.
 
 - [x] My change is adequately tested.
 
   Remember when considering testing:
    + Your change needs to be specifically covered by tests.
       + Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
    + You also need to make sure that the _entire area being changed_ has adequate test coverage.
       + If existing tests don't actually cover the entire area being changed, add tests.
       + This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
    + We should lean on the bulk of code being covered by unit tests, but...
    + ... an end-to-end test should cover the integration points
 
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.

### CODE REVIEW

File: cmd/entrypoint/entrypoint.go [hunk 1]
```diff
@@ -102,6 +102,10 @@ func Main(ctx context.Context, Version string, args ...string) error {
 		// Demo mode!
 		dlog.Infof(ctx, "DEMO MODE")
 		demoMode = true
+		// Set an environment variable so that other parts of the code can check if demo mode is active (mainly used for disabling synthetic authservice injection)
+		os.Setenv("AMBASSADOR_DEMO_MODE", "true")
```
Comment (LanceEa) [inline cmd/entrypoint/entrypoint.go:106]: kind of nitpick/philosophical comment, I don't have too strong of opinions on this but it feels like we have env variable strings spread throughout the code base which makes it difficult to get a full list of what they are (short of hoping docs are up-to-date). 

This is a good example where at a minimum it might be good to bubble it up to the top of file so its visible to the package and can be used within the package.

`const envAmbassadorDemoMode string = "AMBASSADOR_DEMO_MODE"`.

Either prefix or postfix it, maybe as a team we can discuss preferences. The likely chance of it changing is low but this would make it easier in the future to change once (dry principle). The trade-off can be that you have to follow the variable to see the actual string...

File: cmd/entrypoint/env.go [hunk 2]
```diff
@@ -264,6 +265,19 @@ func IsEdgeStack() (bool, error) {
 	}
 }
 
+func IsDemoMode() (bool, error) {
+	demoStr := os.Getenv("AMBASSADOR_DEMO_MODE")
+	if demoStr == "" {
+		demoStr = "false"
```
Comment (LanceEa) [inline cmd/entrypoint/env.go:271]: Why not just bail out early? `return false, nil`

File: cmd/entrypoint/env.go [hunk 3]
```diff
@@ -264,6 +265,19 @@ func IsEdgeStack() (bool, error) {
...
+	}
+	demoMode, err := strconv.ParseBool(demoStr)
+	if err != nil {
+		return false, err
+	} else {
```
Comment (LanceEa) [inline cmd/entrypoint/env.go:276]: else statement is redundant... cleaner just to `return demoMode, nil`

**Review by LanceEa — CHANGES_REQUESTED** [review]
Alice, It looks pretty good. 

I have left some comments. In general, it would be nice to see more unit test for these use cases. Unless I missed it there doesn't seem to be good test coverage for the changes you are making, nor the previous usage of the synthetic function.

File: cmd/entrypoint/env.go [hunk 4]
```diff
@@ -264,6 +265,19 @@ func IsEdgeStack() (bool, error) {
 	}
 }
 
+func IsDemoMode() (bool, error) {
```
[the remark this answers is not in the document]
Comment (AliceProxy) [reply cmd/entrypoint/env.go:268]: Couple notes for this comment:
- We dont have any such package and I didnt feel like creating one to house one function that would be used in one place.
- Making it non-failable makes sense, but this isnt something intended for the user to configure or not. We currently enable demo mode by passing `--demo` at the start, so this environment variable is just for internal use else { to check if demo mode is running or not. Even if a user did set it, the value would be overwritten on startup.

[the remark this answers is not in the document]
Comment (AliceProxy) [reply cmd/entrypoint/env.go:268]: JK, Flynn says we can move this into pkg/dlib :)

File: cmd/entrypoint/syntheticauth.go [hunk 5]
```diff
@@ -40,13 +43,70 @@ func ReconcileAuthServices(ctx context.Context, sh *SnapshotHolder, deltas *[]*k
...
 	}
-	if len(authServices) == 0 && !syntheticAuthExists {
+
+	// Also loop over the annotations and remove authservices that are not v3. We do
+	// this by looping over each entry in the annotations map, removing all the non-v3
+	// AuthService entries, and then removing any keys that end up with empty lists.
+	//
+	// keysToDelete tracks keys to delete, since I never trust deleting things from
+	// collections while iterating over them.
+	keysToDelete := make([]string, 0)
```
[the remark this answers is not in the document]
Comment (AliceProxy) [reply cmd/entrypoint/syntheticauth.go:70]: @kflynn wrote this section that loops through the annotations, and then I edited the commit while cleaning up the git history. Originally he wrote it when we were thinking that these Annotations had to be treated as `kates.Unstructured`. I noticed that they were actually typed so there is some unneeded confusion since I left the comments as they were. I'll clean this section up, but thats just some context for some of the other replies I'll be leaving to provide some info/context before I push some requested changes.

As for deleting items from a collection while iterating over them, I assume Flynn just wrote it that way to be safe with deleting items while iterating over them since we do put a lock on the snapshot in `entrypoint/watcher.go K8sUpdate()` where `ReconcileAuthServices()` is called. There will not be any changes to the snapshot during the course of the function regardless of what happens to the cluster. 

Even if there were rapid config changes in the cluster, we completely finish one update check and complete the reconfigure before begnning another and they all get queued up so nothing is going to be changing in the snapshot. That said, I can still add a test for this to ensure that things are always this way, but I'm just going modify this to delete the keys during the iteration since its safe to do so in golang.

File: cmd/entrypoint/syntheticauth.go [hunk 6]
```diff
@@ -40,13 +43,70 @@ func ReconcileAuthServices(ctx context.Context, sh *SnapshotHolder, deltas *[]*k
...
+				// in our list, we needn't edit it at all. Something to circle back to
+				// later.
+				editedList = append(editedList, annotationObj)
+			}
+		}
+
+		// Once here, is our editedList is empty?
+		if len(editedList) == 0 {
+			// Yes. Delete the whole key for this list.
+			keysToDelete = append(keysToDelete, key)
```
[the remark this answers is not in the document]
Comment (AliceProxy) [reply cmd/entrypoint/syntheticauth.go:105]: The goal is to remove all AuthServices that do not have `protocol_Version: v3` if doing that removal leaves us with an empty list and we assign that empty list back to the key, golang GC to my knowledge will not do anything remove that entry. Then we have a key with an empty list, so we just delete it since theres no point in leaving it as an empty list.

File: cmd/entrypoint/syntheticauth.go [hunk 7]
```diff
@@ -40,13 +43,70 @@ func ReconcileAuthServices(ctx context.Context, sh *SnapshotHolder, deltas *[]*k
...
+		for _, obj := range list {
+			switch annotationObj := obj.(type) {
+			case *v3alpha1.AuthService:
+				// This _is_ an AuthService, so we'll check its protocol version.
+				// Anything other than v3 gets tossed.
+				//
+				// Note also that AuthServices other than getambassador.io/v3 cannot
+				// have a protocol_version, but whatever -- that'll be something not
+				// equal to v3, so it'll get tossed, and that's what we want.
+				if annotationObj.Spec.ProtocolVersion == "v3" {
```
[the remark this answers is not in the document]
Comment (AliceProxy) [reply cmd/entrypoint/syntheticauth.go:86]: re: my point earlier about these comments referring to when this was assuming annotations were `Unstructured`. I think this used to just be something along the lines of:

```
if annotationObj.Object["kind"] == "AuthService" {
  if annotationOb.Object["spec"]["protocol_version"] == "v3" { .......
}
```

where dealing with the object fields is a little less certain than the typeswitch where this is a typed thing that will absolutely have a `protocol_version` field. I can throw a few more cases into the `testutil_fake_syntheticauth_test.go` file to test other apiversions and `protocol_version` values.

File: cmd/entrypoint/syntheticauth.go [hunk 8]
```diff
@@ -40,13 +43,70 @@ func ReconcileAuthServices(ctx context.Context, sh *SnapshotHolder, deltas *[]*k
...
+					// Whoa, it's a v3! Keep it.
+					editedList = append(editedList, annotationObj)
+					authServices = append(authServices, annotationObj)
+					injectSyntheticAuth = false
+				}
+			default:
+				// This isn't an AuthService at all, so we'll keep it.
+				//
+				// XXX There's optimization to do here: if there are no AuthServices
```
[the remark this answers is not in the document]
Comment (AliceProxy) [reply cmd/entrypoint/syntheticauth.go:95]: The only optimization I see here is that we could check if the list contains any AuthServices at all before allocating the memory to copy all the annotations. This is probably preferable to copying it all every time in cases where a user has a ton of annotation config since iterating over a few thousand items is fast, but loading a copy of all those items uses memory if we dont need to in the event that there are no AuthServices.

With that change...
Best case: There are a ton of Annotations, we iterate over them and find there are no AuthServices, then dont do any copying or editing. 

Worst case: There are a ton of Annotations, we iterate over them and find there is an AuthService we need to drop, but we have to iterate over n-1 items in the list of annotations to find it, then we iterate over it again but this time making a copy of all the resources but leave out that AuthService we want to drop.

The small amount of extra time it takes to iterate is probably worth saving the memory, but a user with this much annotation config is an edge-case and also the memory saved would be small. I'll commit a change and you can let me know if you think the tradeoff is worth it or a waste.

File: cmd/entrypoint/entrypoint.go (hunk 1 above)
Comment (AliceProxy) [reply cmd/entrypoint/entrypoint.go:106]: Sure, this makes sense and I think making it more visible by bubling it up to the top of the file (and prefixing it with env... will make things more searchable to look for all the environment variables). I'd say definitely worth the tradeoff of having to check the variable.

File: cmd/entrypoint/syntheticauth.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (LanceEa) [reply cmd/entrypoint/syntheticauth.go]: ok, yes that makes sense. I'm guessing the memory savings are minor and your right GC wouldn't release the empty list at that point.

**Review by LanceEa — APPROVED** [review]
lgtm
